src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Configuration;
+using NuGet.Protocol.Core.Types;
+using NuGet.Protocol.Plugins;
+using System.Linq;
+
+namespace NuGet.Credentials
+{
+    public class SecurePluginCredentialProvider : ICredentialProvider
+    {
+
+        private PluginDiscoveryResult DiscoveredPlugin { get; set; }
+        private Common.ILogger Logger { get; set; }
+
+        // We use this to avoid needlessly instantiating plugins if they don't support authentication.
+        private bool IsAnAuthenticationPlugin { get; set; } = true;
+
+        public SecurePluginCredentialProvider(PluginDiscoveryResult pluginDiscoveryResult, Common.ILogger logger)
+        {
+            if (pluginDiscoveryResult == null)
+            {
+                throw new ArgumentNullException(nameof(pluginDiscoveryResult));
+            }
+            if (pluginDiscoveryResult.PluginFile.State != PluginFileState.Valid)
+            {
+                throw new ArgumentException("Cannot create a provider from an invalid plugin. Plugin state: " + pluginDiscoveryResult.PluginFile.State + " " + pluginDiscoveryResult.Message);
```

Comment:
Add an argument to `ArgumentException` constructor:  `nameof(pluginDiscoveryResult)`